fix(test-types): format `Transaction` values as readable hex-strings instead of raw bytes (#1803)

felix314159 opened the pull request:
## 🗒️ Description
```
a = Transaction()
print(f"{a}")
```
Before this PR the resulting log is unreadable, you don't even see who is sending to who. With this PR you get readable addresses 0x...

## 🔗 Related Issues or PRs
<!-- Reference any related issues using the GitHub issue number (e.g., Fixes #123). Default is N/A. -->
N/A.

## ✅ Checklist
<!-- Please check off all required items. For those that don't apply remove them accordingly. -->

- [ ] All: Ran fast `tox` checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    uvx tox -e static
    ```
- [ ] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.
- [ ] All: Considered adding an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [ ] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [ ] All: Set appropriate labels for the changes (only maintainers can apply labels).
- [ ] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://eest.ethereum.org/main/tests/) are correctly formatted.
- [ ] Tests: For PRs implementing a missed test case, update the [post-mortem document](/ethereum/execution-spec-tests/blob/main/docs/writing_tests/post_mortems.md) to add an entry the list.
- [ ] Ported Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) or [tests/static](/ethereum/execution-spec-tests/blob/main/tests/static) have been assigned `@ported_from` marker.

#### Cute Animal Picture
<img src="https://cdn.discordapp.com/attachments/1337018504141864963/1460628488066171106/IMG_0019.jpg?ex=69679bc8&is=69664a48&hm=2dd5f39fece13bf15a305553e507efb8a18b7075c1e36d95705a5a3190cc207d&" width="50%" alt="Description">

fselmo commented on packages/testing/src/execution_testing/test_types/transaction_types.py:
```diff
@@ -730,6 +725,42 @@ def metadata_string(self) -> str | None:
+
+        Special handling for to, data, and sender fields to show hex encoding.
+        """
+        if field_name in ("to", "data", "sender"):
```
What do you think about not being restrictive to specific fields here? Something like:

```py
def _format_field_value(self, value: Any) -> str:
    """
    Format a field value for string representation.
    
    Uses hex encoding for any value that supports it
    (Address, Bytes, Hash, HexNumber, etc).
    """
    if value is None:
       return "None"
    elif hasattr(value, "hex"):
       return value.hex()
    else:
       return repr(value)
```

I did a quick test and this works well. If we go with this, we just need to update the docstring descriptions in `__repr__()` and `__str__()`.

felix314159 replied:
Fine with me. If we ever add sth like a `float` field to `Transaction` we can add an exception, for some datatypes it looks ugly e.g.:
```
0.1.hex()
'0x1.999999999999ap-4'
```

fselmo commented on packages/testing/src/execution_testing/test_types/transaction_types.py:
```diff
@@ -730,6 +725,42 @@ def metadata_string(self) -> str | None:
+            formatted_value = self._format_field_value(field_name, value)
+            field_strs.append(f"{field_name}={formatted_value}")
+
+        return " ".join(field_strs)
```
nit: is there a good reason to deviate from the usual pattern for `__repr__()` here with `ClassName(...)`?

```suggestion
        return f"{self.__class__.__name__}({', '.join(field_strs)})"
```

felix314159 replied:
Thanks I prefer your approach, it makes object boundaries more explicit. E.g. we might log an object whose fields include other objects and then it's nice that there is clear separation between the parent's fields and the nested object's representation. e.g.
```
child=a=1 b=2 status='ok'
```
vs
```
ParentWrapped(child=ChildWrapped(a=1, b=2), status='ok')
```
Thanks good change, I actually prefer being explicit here cuz it helps sometimes when debugging

fselmo reviewed (commented):
I support better readability 👌🏼. Just a minor suggestion for generalizing it to other fields with hex support. Can do another pass after this is rebased.

felix314159 committed:
implemented fselmo feedback + replaced deprecated  with

felix314159 commented:
TIL:
git commit -m "abc \`def\` ghi" shows up as "abc ghi"
but
git commit -m 'abc \`def\` ghi' shows up as "abc `def` ghi`

fselmo commented on packages/testing/src/execution_testing/test_types/transaction_types.py:
```diff
@@ -842,6 +842,37 @@ def signer_minimum_balance(self, *, fork: Fork) -> int:
+        if value is None:
+            return "None"
+        elif hasattr(value, "hex"):
+            return value.hex()
```
Ah my fault on the currently failing tests. I think my code snippet was a bit off. The `__repr__` is generating an invalid hex when `data=b""` (empty bytes) because `Bytes(b"").hex()` returns `"0x"` which is an invalid hex (it needs to be something like `0x0` with one byte at least).

We can change this to the string representation directly and it should be OK:

```suggestion
            return f'"{value.hex()}"'
```

danceratopz replied:
Yes, thanks @fselmo! This was necessary to fix the remaining failing tests!

danceratopz commented:
@felix314159 just rebased on main to get the potential fix from:
- #2009

danceratopz committed:
fix(gentest): improve ruff error reporting with full diagnostic output

Remove --quiet flag and include returncode, stdout, and stderr in
error messages to enable diagnosis of formatting failures.

danceratopz committed:
fix(test-types): ensure valid hex field values are returned

Wrap hex values in quotes to produce valid Python string literals
in Transaction.__repr__().

Previously, empty bytes (data=b"") would generate `data=0x` via
Bytes(b"").hex() returning "0x", which is invalid Python syntax
(needs at least one digit like 0x0). This caused ruff to fail when
formatting generated test code.

Now hex values are quoted strings (e.g., data="0x") which are always
valid Python and accepted by the framework's type coercion.

Co-authored-by: @fselmo

danceratopz commented on packages/testing/src/execution_testing/cli/gentest/source_code_generator.py:
```diff
@@ -100,7 +99,8 @@ def format_code(code: str) -> str:
             raise Exception(
                 f"Error formatting code using formatter '{formatter_path}': "
-                f"{result.stderr}"
+                f"returncode={result.returncode}, stdout={result.stdout!r}, "
```
Unrelated change, just an improvement to help debug the issues we were seeing in CI.

danceratopz reviewed (commented):
[capture_gentest.py](https://github.com/user-attachments/files/24587394/capture_gentest.py)

Thanks for adding this @felix314159! The CI fails were not due to a race condition while executing the gentest tests via `CliRunner`. `gentest` formats the generated code with `ruff`, which was failing because it generated invalid Python code until the fix from @fselmo was applied (`f'"{value.hex()}"'` instead of `value.hex()`, to ensure `0x0`, not `0x` was returned).

After seeing this fail (which is now fixed), I thought wed better check what else has changed with the output of `gentest` and these changes do greatly improve situation for `Address` objects (which was a long time coming, so thanks!), but actually regresses the situation for numeric values, such as `gas_price`. The following summarizes the situation.

The `_format_field_value` change improves readability for addresses and bytes data, but reduces readability for numeric fields.

## Examples from gentest output

**Addresses/Data (improved):**
```python
# Before: raw bytes (hard to read)
to=b'\xc6\xd9g\x86G\x7f\x82I\x1b\xfe\xad\x8f\x00\xb8)F\x88\xf7z\xbc'
data=b''

# After: hex strings (readable)
to="0xc6d96786477f82491bfead8f00b8294688f77abc"
data="0x"
```

**Numeric fields (regressed):**
```python
# Before: decimal (clear)
gas_price=10000000000  # obviously 10 gwei
value=668250000000000000
nonce=2

# After: hex strings (opaque)
gas_price="0x2540be400"  # what's this in gwei?
value="0x94619d59f27a000"
nonce="0x2"
```

## Suggestion

Consider making `_format_field_value` more nuanced:
- Addresses/Bytes → use hex string (current behavior, good)
- Integers/HexNumber → keep as decimal repr (better for readability)

This could be done by checking the type more specifically rather than just `hasattr(value, "hex")`.

I attached the script that Claude wrote to help compare the output.

felix314159 committed:
show int and other numbers as decimal

felix314159 commented:
Sorry I thought I had the SLOTNUM PR open when changing the target branch here :D

danceratopz commented on packages/testing/src/execution_testing/test_types/transaction_types.py:
```diff
@@ -842,6 +844,48 @@ def signer_minimum_balance(self, *, fork: Fork) -> int:
+                # while removing unnecessary trailing zeros (100.000000 -> 100)
+                return "{:f}".format(value).rstrip("0").rstrip(".")
+
+            return str(value)
```
I re-ran the script, we need this additional change to get decimal output, otherwise:
- nonce=0x2
- gas_price=0x2540be400
- value=0x94619d59f27a000

```suggestion
            # Force decimal representation for int subclasses like HexNumber
            if isinstance(value, int):
                return str(int(value))

            return str(value)
```

fselmo replied:
Good catch!

danceratopz commented on packages/testing/src/execution_testing/test_types/transaction_types.py:
```diff
@@ -842,6 +844,48 @@ def signer_minimum_balance(self, *, fork: Fork) -> int:
+        Format a field value for string representation.
+
+        Uses hex encoding for any value that supports it
+        (Address, Bytes, Hash, HexNumber, etc).
```
```suggestion
        Uses decimal for numeric values (int, HexNumber, etc.) and
        hex encoding for Address, Bytes, Hash, etc.
```

danceratopz commented on packages/testing/src/execution_testing/test_types/transaction_types.py:
```diff
@@ -842,6 +844,48 @@ def signer_minimum_balance(self, *, fork: Fork) -> int:
+                # Convert to string to avoid scientific notation (1E-9)
+                # while removing unnecessary trailing zeros (100.000000 -> 100)
+                return "{:f}".format(value).rstrip("0").rstrip(".")
+
```
This seems overly defensive? We don't use Decimal anywhere in `package/testing/src/`.
```suggestion
```

danceratopz reviewed (commented):
Thanks @felix314159, here's few comments on the last change.

felix314159 committed:
Update packages/testing/src/execution_testing/test_types/transaction_types.py

Co-authored-by: danceratopz <danceratopz@gmail.com>

felix314159 committed:
Update packages/testing/src/execution_testing/test_types/transaction_types.py

Co-authored-by: danceratopz <danceratopz@gmail.com>

felix314159 committed:
Update packages/testing/src/execution_testing/test_types/transaction_types.py

Co-authored-by: danceratopz <danceratopz@gmail.com>

danceratopz committed:
style(test-types): remove unused import, fix ruff

danceratopz reviewed (approved):
Thanks for this improvement, @felix314159! LGTM!

fselmo reviewed (approved):
You've got my approval as well, thanks for this! Let's just make sure to `squash + merge` 😅.
